File: combinatorics/resource/LDL/Python/bai3.py
def is_self_conjugate(partition):
    max_len = max(partition)
    conjugate = []

    for i in range(1, max_len + 1):
        count = 0
        for val in partition:
            if val >= i:
                count += 1
            else:
                break
        conjugate.append(count)

    return conjugate == partition


def generate_self_conjugate_partitions(n, k=None):
    result = []

    def backtrack(remaining, curr, max_val):
        if remaining == 0:
            if k is None or len(curr) == k:
                sorted_part = sorted(curr, reverse=True)
                if is_self_conjugate(sorted_part):
                    result.append(sorted_part)
            return
        for i in range(min(remaining, max_val), 0, -1):
            curr.append(i)
            backtrack(remaining - i, curr, i)
            curr.pop()

    backtrack(n, [], n)
    return result


def generate_odd_partitions(n):
    result = []

    def backtrack(remaining, curr, min_odd):
        if remaining == 0:
            result.append(curr.copy())
            return
        for i in range(min_odd, remaining + 1, 2):
            curr.append(i)
            backtrack(remaining - i, curr, i)
            curr.pop()

    backtrack(n, [], 1)
    return result


# (c.i) De quy
def p_k_selfconj_recursive(n, k, memo={}):
    key = (n, k)
    if key in memo:
        return memo[key]
    if n == 0 and k == 0:
        return 1
    if n <= 0 or k <= 0:
        return 0

    count = 0

    def backtrack(remaining, parts_left, curr, max_val):
        nonlocal count
        if remaining == 0 and parts_left == 0:
            sorted_part = sorted(curr, reverse=True)
            if is_self_conjugate(sorted_part):
                count += 1
            return
        if remaining <= 0 or parts_left <= 0:
            return
        for i in range(min(remaining, max_val), 0, -1):
            curr.append(i)
            backtrack(remaining - i, parts_left - 1, curr, i)
            curr.pop()

    backtrack(n, k, [], n)
    memo[key] = count
    return count


# (c.ii) Quy hoach dong
def p_k_selfconj_dp(n, k):
    dp = [[0] * (k + 1) for _ in range(n + 1)]
    dp[0][0] = 1

    for sum_val in range(1, n + 1):
        for parts in range(1, min(k, sum_val) + 1):
            count = 0

            def backtrack(remaining, parts_left, curr, max_val):
                nonlocal count
                if remaining == 0 and parts_left == 0:
                    sorted_part = sorted(curr, reverse=True)
                    if is_self_conjugate(sorted_part):
                        count += 1
                    return
                if remaining <= 0 or parts_left <= 0:
                    return
                for i in range(min(remaining, max_val), 0, -1):
                    curr.append(i)
                    backtrack(remaining - i, parts_left - 1, curr, i)
                    curr.pop()

            backtrack(sum_val, parts, [], sum_val)
            dp[sum_val][parts] = count

    return dp[n][k]

if __name__ == "__main__":
    n = int(input("Nhap n (so can phan hoach): "))
    k = int(input("Nhap k (so phan cua tu lien hop): "))

    # (a) Dem va in cac phan hoach tu lien hop co k phan
    print(f"\n(a) Cac phan hoach tu lien hop cua {n} co {k} phan:")
    sc_parts = generate_self_conjugate_partitions(n, k)
    for p in sc_parts:
        print("  ", p)
    print(f"=> So phan hoach tu lien hop p_k^selfcjg({n},{k}) = {len(sc_parts)}")

    # (b) So sanh voi so phan hoach co so luong phan tu la so le
    odd_parts = generate_odd_partitions(n)
    print(f"\n(b) So phan hoach cua {n} co so luong phan la so le:")
    for p in odd_parts:
        print("  ", p)
    print(f"=> Tong cong co {len(odd_parts)} phan hoach co so phan le")

    # (c.i) De quy
    res_recursive = p_k_selfconj_recursive(n, k)
    print(f"\n(c.i) Tinh bang de quy: p_k^selfcjg({n},{k}) = {res_recursive}")

    # (c.ii) Quy hoach dong
    res_dp = p_k_selfconj_dp(n, k)
    print(f"(c.ii) Tinh bang quy hoach dong: p_k^selfcjg({n},{k}) = {res_dp}")

File: combinatorics/resource/LDL/C++/bai3.cpp
#include <bits/stdc++.h>
using namespace std;

// Ham kiem tra mot phan hoach co phai la tu lien hop hay khong
bool is_self_conjugate(const vector<int> &partition) {
    if (partition.empty()) return false;
    int max_len = partition[0]; // vi partition da sap xep giam dan
    vector<int> conjugate;
    for (int i = 1; i <= max_len; i++) {
        int count = 0;
        for (int val : partition) {
            if (val >= i)
                count++;
            else
                break;
        }
        conjugate.push_back(count);
    }
    return conjugate == partition;
}

// Sinh cac phan hoach tu lien hop cua n, neu k != -1 thi chi lay cac phan co k phan tu
void backtrack_self_conjugate(int remaining, vector<int> &curr, int max_val,
                              int n, int k, vector<vector<int>> &result) {
    if (remaining == 0) {
        if ((k == -1 || (int)curr.size() == k)) {
            vector<int> sorted_part = curr;
            sort(sorted_part.rbegin(), sorted_part.rend());
            if (is_self_conjugate(sorted_part)) {
                result.push_back(sorted_part);
            }
        }
        return;
    }
    for (int i = min(remaining, max_val); i >= 1; --i) {
        curr.push_back(i);
        backtrack_self_conjugate(remaining - i, curr, i, n, k, result);
        curr.pop_back();
    }
}

vector<vector<int>> generate_self_conjugate_partitions(int n, int k = -1) {
    vector<vector<int>> result;
    vector<int> curr;
    backtrack_self_conjugate(n, curr, n, n, k, result);
    return result;
}

// Sinh cac phan hoach cua n chi bao gom so le
void backtrack_odd(int remaining, vector<int> &curr, int min_odd,
                   vector<vector<int>> &result) {
    if (remaining == 0) {
        result.push_back(curr);
        return;
    }
    for (int i = min_odd; i <= remaining; i += 2) {
        curr.push_back(i);
        backtrack_odd(remaining - i, curr, i, result);
        curr.pop_back();
    }
}

vector<vector<int>> generate_odd_partitions(int n) {
    vector<vector<int>> result;
    vector<int> curr;
    backtrack_odd(n, curr, 1, result);
    return result;
}

// (c.i) De quy co nho
int p_k_selfconj_recursive(int n, int k, map<pair<int, int>, int> &memo) {
    pair<int, int> key = {n, k};
    if (memo.count(key)) return memo[key];
    if (n == 0 && k == 0) return 1;
    if (n <= 0 || k <= 0) return 0;

    int count = 0;
    vector<int> curr;

    function<void(int, int, vector<int> &, int)> backtrack = [&](int remaining, int parts_left, vector<int> &curr, int max_val) {
        if (remaining == 0 && parts_left == 0) {
            vector<int> sorted_part = curr;
            sort(sorted_part.rbegin(), sorted_part.rend());
            if (is_self_conjugate(sorted_part))
                count++;
            return;
        }
        if (remaining <= 0 || parts_left <= 0) return;
        for (int i = min(remaining, max_val); i >= 1; --i) {
            curr.push_back(i);
            backtrack(remaining - i, parts_left - 1, curr, i);
            curr.pop_back();
        }
    };

    backtrack(n, k, curr, n);
    memo[key] = count;
    return count;
}

// (c.ii) Quy hoach dong: Kiem tra tu lien hop cho tung phan hoach sinh ra
int p_k_selfconj_dp(int n, int k) {
    vector<vector<int>> dp(n + 1, vector<int>(k + 1, 0));
    dp[0][0] = 1;
    for (int sum_val = 1; sum_val <= n; ++sum_val) {
        for (int parts = 1; parts <= min(k, sum_val); ++parts) {
            int count = 0;
            vector<int> curr;

            function<void(int, int, vector<int> &, int)> backtrack = [&](int remaining, int parts_left, vector<int> &curr, int max_val) {
                if (remaining == 0 && parts_left == 0) {
                    vector<int> sorted_part = curr;
                    sort(sorted_part.rbegin(), sorted_part.rend());
                    if (is_self_conjugate(sorted_part))
                        count++;
                    return;
                }
                if (remaining <= 0 || parts_left <= 0) return;
                for (int i = min(remaining, max_val); i >= 1; --i) {
                    curr.push_back(i);
                    backtrack(remaining - i, parts_left - 1, curr, i);
                    curr.pop_back();
                }
            };

            backtrack(sum_val, parts, curr, sum_val);
            dp[sum_val][parts] = count;
        }
    }
    return dp[n][k];
}

int main() {
    int n, k;
    cout << "Nhap n (so can phan hoach): ";
    cin >> n;
    cout << "Nhap k (so phan cua tu lien hop): ";
    cin >> k;

    // (a) Dem va in cac phan hoach tu lien hop co k phan
    cout << "\n(a) Cac phan hoach tu lien hop cua " << n << " co " << k << " phan:\n";
    auto sc_parts = generate_self_conjugate_partitions(n, k);
    for (auto &p : sc_parts) {
        cout << "  ";
        for (int x : p) cout << x << " ";
        cout << "\n";
    }
    cout << "=> So phan hoach tu lien hop p_k^selfcjg(" << n << "," << k << ") = " << sc_parts.size() << "\n";

    // (b) So sanh voi phan hoach co cac so le
    auto odd_parts = generate_odd_partitions(n);
    cout << "\n(b) So phan hoach cua " << n << " co cac phan tu la so le:\n";
    for (auto &p : odd_parts) {
        cout << "  ";
        for (int x : p) cout << x << " ";
        cout << "\n";
    }
    cout << "=> Tong cong co " << odd_parts.size() << " phan hoach co so phan tu la so le\n";

    // (c.i) De quy
    map<pair<int, int>, int> memo;
    int res_recursive = p_k_selfconj_recursive(n, k, memo);
    cout << "\n(c.i) Tinh bang de quy: p_k^selfcjg(" << n << "," << k << ") = " << res_recursive << "\n";

    // (c.ii) Quy hoach dong
    int res_dp = p_k_selfconj_dp(n, k);
    cout << "(c.ii) Tinh bang quy hoach dong: p_k^selfcjg(" << n << "," << k << ") = " << res_dp << "\n";

    return 0;
}